### PR TITLE
Add file selection mode

### PR DESCRIPTION
**Pull request by DaveJarvis**
Adds ability to limit the types of files that can be selected.

Production changes:

* **JWFileChooserViewController.java** -- Add new "filterSelectionMode" method that answers whether a given `Path` may be accepted for selection and return.
* **JWFileChooser.java** -- Remove trivial code duplication.

Unit test changes:

* **JWFileChooserDefaultModeTest.java** -- Verifies that the default behaviour remains the same. We could consider removing this test as it may be covered by the other tests.
* **JWFileChooserDirectoryModeTest.java** -- Verifies that restricting the selection mode to a particular file (in this case, the "DOC" directory) prevents users from selecting other files without introducing inconsistent behaviour.
* **JWRobotDelegate.java** -- Isolates common, reusable behaviours, to simplify unit tests.

### CODE REVIEW

**Comment by io7m** [comment]
I think this looks good in general, and I appreciate the updates to the test suite. You have some style checking issues though (mostly around star imports). Once those are fixed, I'm happy to merge this in, and I'll go through the rest of the test suite and use the delegate there.

**Comment by io7m** [comment]
Our CI actions record a video of the test suite now, in case you've not noticed. It's kind of fun! :smile:

**Comment by DaveJarvis** [comment]
`JWFileChooserFilterDefaultTest::testFilterDefaultHidesEverything` works when run in the IDE, but fails when running `mvn clean package`. Is there a race condition?

Checkstyle incorrectly reports on a valid Javadoc usage. Javadocs will be generated correctly when referencing an imported class via the `{@link ...}` syntax. For example, in `JWFileChooserConfigurationType`, I wanted to write:

>     /**
>      * Sets a mode that prevents returning a file to the client if that file
>      * does not meet criteria returned by this method. For example, this may be
>      * used to prevent selecting directories by implementing a function that
>      * returns {@code true} when {@link Files#isRegularFile(Path, LinkOption...)}

But Checkstyle cannot resolve the `Files` reference, despite it being imported. Consider disabling that particular check.

**Comment by DaveJarvis** [comment]
You may wish to take a closer look at `getTableColumn`: I have not verified that it is correct.

**Comment by io7m** [comment]
> JWFileChooserFilterDefaultTest::testFilterDefaultHidesEverything works when run in the IDE, but fails when running mvn clean package. Is there a race condition?

There could well be. I'll look into it!

**Comment by io7m** [comment]
> But Checkstyle cannot resolve the Files reference, despite it being imported. Consider disabling that particular check.

This seems to be a sporadic bug in Checkstyle. It's possible to circumvent it by using the fully-qualified name in the comment:

```
{@link java.nio.file.Files#isRegularFile(Path, LinkOption...)}
```

File: com.io7m.jwheatsheaf.tests/src/test/java/com/io7m/jwheatsheaf/tests/JWRobotDelegate.java [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+
+  TableCell<?, ?> getTableCellFileName( final String text ) {
+    return
+      this.robot.lookup( n -> n instanceof TableCell )
+                .queryAllAs( TableCell.class )
+                .stream()
+                .filter( cell -> Objects.equals( cell.getText(), text ) )
+                .findFirst()
+                .orElseThrow( () -> new IllegalStateException(
+                  "Unable to locate a '.' directory entry" )
```
Comment (DaveJarvis) [inline com.io7m.jwheatsheaf.tests/src/test/java/com/io7m/jwheatsheaf/tests/JWRobotDelegate.java:64]: The `'.'` should be `text`.

**Comment by io7m** [comment]
I'll be attempting to merge this in over the next few days. At the very least, I'd like to get the robot delegate used everywhere. :smile:

**Comment by io7m** [comment]
There's some slightly weird behaviour here with regards to selecting the entry ".". I'm looking into it now.

**Comment by io7m** [comment]
There's an interaction between this and some code that hit `develop` before you even wrote this code. I'm sorting it out now, it's easy to fix.

**Pull request merged**